### DISCUSSION TITLE
GitHub Repo Search using tabs and recent database

### DISCUSSION THREAD

**Discussion by elijahmanor** [johnlindquist/kit#75] [Q&A] [answered]
First of all, I love the Kit project @johnlindquist!

I've started to leverage kit to create a GitHub Repo Serach script and started to use the new Tab feature.

![Kapture 2021-03-24 at 16 10 29](https://user-images.githubusercontent.com/86454/112383729-80a4f780-8cbb-11eb-9901-0e8941182147.gif)

<details>
<summary>Click to expand code</summary>
<p>

#### github-repo.js

```js
// Menu: GitHub Repository Search
// Description: Search for a GitHub Repository
// Author: Elijah Manor
// Twitter: @elijahmanor

let { setSelectedText } = await kit("text");
let scriptDB = db("github-repo", { history: [] });

const open = (url) => {
  exec(`open ${url}`);
  setSelectedText(url);
};

onTab("Search", async () => {
  let config = {
    headers: {
      Authorization:
        "Bearer " +
        (await env("GITHUB_REPO_TOKEN", {
          info: `Create a GitHub Repo token: <a class="bg-white" href="https://github.com/settings/tokens/new">https://github.com/settings/tokens/new</a>`,
          message: `Set .env GITHUB_REPO_TOKEN:`,
        })),
    },
  };

  let lastSearch = [];
  let repositories = async (query) => {
    if (query.length < 4) return [];
    const response = await get(
      `https://api.github.com/search/repositories?q=${query}`,
      config
    );
    lastSearch = response.data.items.map((repo) => {
      return {
        name: repo.full_name,
        value: repo.html_url,
        description: repo.description,
        html: `<img class="h-12 w-12 flex-none object-contain" src="${repo.owner.avatar_url}" />`,
      };
    });
    return lastSearch;
  };
  let repositoryUrl = await arg("Search for repository", repositories);

  open(repositoryUrl);

  let recentHistory = scriptDB.get("history").value();

  let index = recentHistory.findIndex((r) => r.value === repositoryUrl);
  recentHistory = [
    lastSearch.find((r) => r.value === repositoryUrl),
    ...recentHistory.slice(0, index),
    ...recentHistory.slice(index + 1),
  ];

  scriptDB.set("history", recentHistory).write();
});

onTab("Recent", async () => {
  let recentHistory = scriptDB.get("history").value();
  let repositoryUrl = await arg("Select recent repository", recentHistory);
  open(repositoryUrl);
});
```

</p>
</details>

Everything seems to be working great, but I'm finding the entries that I store in the `db` are bleeding into the main application (outside my script). Am I using the `db` incorrectly or is there an internal bug in Kit? You can see the bug at the end of the GIF.

Since I am supporting live searches (search as you type), I ended up saving off the last search response so that I could access it outside of the async function. Do you know of a better way to do that? The reason I did that was to get access to the full item, not just the value that was returned. If the return was a richer object (not only the value) then that could work as well.

Anyway, love playing with Kit. I already have several scripts I'm playing with.

**Comment by johnlindquist** [comment, marked as answer]
The `value` can be set to anything:

Sidenote: "anything" that's can be json serialized. I just realized there may be cases where you want to return a fn that I haven't considered. I may want the script/app to communicate with choice ids rather than JSON objects 🤔

Back on track. This `value` is set to include the `repo` and the `query`. I've found this situation pop up a few times where I want to continue the script with the user's input as well:

```js
     return {
        name: repo.full_name,
        value: {repo, query},
        description: repo.description,
        html: `<img class="h-12 w-12 flex-none object-contain" src="${repo.owner.avatar_url}" />`,
      };
```

I'll dig into the "db bleeding over" issue today. I _think_ it's an "on submit, clear choices" issue. I've recently told the App to persist choices if you're in the same script context (because of tabs), so I was expecting a few of these cases to pop up. Thanks so much for reporting.

**Reply by elijahmanor** [reply]
Ohh, I didn't think the value could be anything! That solves the problem. I didn't like doing what I was doing (setting memory outside of the function) so this is a great solution. I'll refactor the code to make it cleaner. 

Thanks for looking into the "db bleeding over" issue!

**Reply by johnlindquist** [reply]
To be even more specific, here's the line of code where a choice is selected in the App:

https://github.com/johnlindquist/kitapp/blob/beta/src/kit.ts#L65

The app sends it to the child process (your current running script) which receives it here:

https://github.com/johnlindquist/kit/blob/main/preload/app.cjs#L139

The message handler, which has an open connection to the App, is waiting for any valid data to come back.
